### PR TITLE
Polling

### PR DESCRIPTION
**Pull request by iloveitaly**
Initial polling implementation! Curious to hear your thoughts @AlessioRocco. There are still some loose ends, and some decisions that don't make the implementation completely backwards compatible with Wombat—I wanted to get your feedback before moving forward.

TODO

- [x] PollJob spec
- [x] Use `ClassConfiguration` in `Cangaroo::Job`
- [x] Fix poll timestamp validations
- [x] Look into making parameter value backwards compatible with wombat. It might be easier to force wombat endpoints to use a consistent polling parameter.

### CODE REVIEW

**Comment by AlessioRocco** [comment]
@iloveitaly I did some changes on master to make our life easier. 

I've updated the [Cangaroo::Job](https://github.com/nebulab/cangaroo/blob/master/app/jobs/cangaroo/job.rb#L39-L45) with this [code](https://github.com/iloveitaly/cangaroo/blob/polling/app/jobs/cangaroo/poll_job.rb#L18-L23), this was a Wombat feature that Cangaroo miss, with this feature the flow is restarted in case an Integration responds with some data, I've also created the [Cangaroo::PerformFlow](https://github.com/nebulab/cangaroo/blob/master/app/interactors/cangaroo/perform_flow.rb) organizer because we already have the `destination connection` inside the Job and use `Cangaroo:HandleRequest` create an extra query to find it.
With the new changes you can let `Cangaroo::PollJob` inherit from [Cangaroo::Base](https://github.com/nebulab/cangaroo/blob/master/app/jobs/cangaroo/job.rb) and have less code duplication, you can for example put https://github.com/iloveitaly/cangaroo/blob/polling/app/jobs/cangaroo/poll_job.rb#L13 in the `before_perform` and https://github.com/iloveitaly/cangaroo/blob/polling/app/jobs/cangaroo/poll_job.rb#L25-L27 in the `after_perform`, but these are only suggestions so implement it how do you think it's better.

My idea for the future is to have three classes `Cangaroo::Job`, `Cangaroo::PushJob` and `Cangaroo::PullJob` where `Cangaroo::Job` act as base class for `Cangaroo::PushJob` and `Cangaroo::PullJob` but also for user custom `Jobs` that aren't related to Integrations, but this need a `major` release because isn't backwards compatible. 
To reach this goal I thought that we can first merge the `ClassConfiguration` into `Cangaroo::Job` then create the `Cangaroo::PollJob` that inherit from `Cangaroo::Job` and after refactor the code to have the three classes and update the version to `2.0.0`.

About use a consistent polling parameter for `Cangaroo::PullJob` I really like the idea but seems a big work for me to update all existing Integrations with that, maybe we can let the end developer manage the needed parameters for now so we can release this feature and then work on update all Integrations with a consistent polling parameter.

To recap my thoughts are:

1. Use `ClassConfiguration` in `Cangaroo::Job` (can you please create this PR?)
2. Add `Cangaroo::PollJob`
3. Refactor for `Cangaroo::Job`, `Cangaroo::PushJob` and `Cangaroo::PullJob` then update to version `2.0.0`

What do you think about that? Can it work for you?

**Comment by iloveitaly** [comment]
:+1: Sounds good!

Request 1 is handled in https://github.com/nebulab/cangaroo/pull/17

**Comment by iloveitaly** [comment]
>Look into making parameter value backwards compatible with wombat. It might be easier to force wombat endpoints to use a consistent polling parameter.

@AlessioRocco we chatted about this offline: each wombat endpoint uses a different last poll param. We can't implement this.

---

Updated this PR to contain the changes we talked about. This should be good to go!

**Comment by bricesanchez** [comment]
Any update ?

**Comment by AlessioRocco** [comment]
I had some busy days but I hope to see this and other PR this weekend.

**Comment by bricesanchez** [comment]
Thanks @AlessioRocco !

**Comment by iloveitaly** [comment]
@AlessioRocco I've had this running in production for a couple weeks, and there were a couple changes that came up. Hold off on reviewing this PR until I've had a chance to pull in those bug fixes.

I'll ping you on this PR when I've merged them in.

**Comment by bricesanchez** [comment]
I'm testing the Polling feature and i've this error, when i can a PollJob : 

```ruby
irb(main):012:0> Cangaroo::RunPolls.call(jobs: [Cangaroo::ProductPollJob])
Performing Cangaroo::ProductPollJob from Inline(cangaroo)
  Cangaroo::Connection Load (0.4ms)  SELECT  "cangaroo_connections".* FROM "cangaroo_connections" WHERE "cangaroo_connections"."name" = $1 LIMIT 1  [["name", "chaindrive"]]
  Cangaroo::PollTimestamp Load (0.3ms)  SELECT  "cangaroo_poll_timestamps".* FROM "cangaroo_poll_timestamps" WHERE "cangaroo_poll_timestamps"."job" = $1 AND "cangaroo_poll_timestamps"."connection_id" = 3  ORDER BY "cangaroo_poll_timestamps"."id" ASC LIMIT 1  [["job", "Cangaroo::ProductPollJob"]]
  Cangaroo::Connection Load (0.2ms)  SELECT  "cangaroo_connections".* FROM "cangaroo_connections" WHERE "cangaroo_connections"."name" = $1 LIMIT 1  [["name", "chaindrive"]]
  Cangaroo::PollTimestamp Load (0.4ms)  SELECT  "cangaroo_poll_timestamps".* FROM "cangaroo_poll_timestamps" WHERE "cangaroo_poll_timestamps"."job" = $1 AND "cangaroo_poll_timestamps"."connection_id" = 3  ORDER BY "cangaroo_poll_timestamps"."id" ASC LIMIT 1  [["job", "Cangaroo::ProductPollJob"]]
I, [2016-04-25T16:14:13.251498 #38069]  INFO -- : initiating poll: job=Cangaroo::ProductPollJob job_id=ce701595-a390-45ba-adb4-3eb29a5c638f connection=chaindrive last_poll=0
  Cangaroo::Connection Load (0.2ms)  SELECT  "cangaroo_connections".* FROM "cangaroo_connections" WHERE "cangaroo_connections"."name" = $1 LIMIT 1  [["name", "chaindrive"]]
  Cangaroo::Connection Load (0.2ms)  SELECT  "cangaroo_connections".* FROM "cangaroo_connections" WHERE "cangaroo_connections"."name" = $1 LIMIT 1  [["name", "chaindrive"]]
  Cangaroo::PollTimestamp Load (0.3ms)  SELECT  "cangaroo_poll_timestamps".* FROM "cangaroo_poll_timestamps" WHERE "cangaroo_poll_timestamps"."job" = $1 AND "cangaroo_poll_timestamps"."connection_id" = 3  ORDER BY "cangaroo_poll_timestamps"."id" ASC LIMIT 1  [["job", "Cangaroo::ProductPollJob"]]
I, [2016-04-25T16:14:16.986170 #38069]  INFO -- : processing poll results: job=Cangaroo::ProductPollJob job_id=ce701595-a390-45ba-adb4-3eb29a5c638f connection=chaindrive
  Cangaroo::Connection Load (0.4ms)  SELECT  "cangaroo_connections".* FROM "cangaroo_connections" WHERE "cangaroo_connections"."key" = $1 AND "cangaroo_connections"."token" = $2  ORDER BY "cangaroo_connections"."id" ASC LIMIT 1  [["key", "123"], ["token", "123"]]
Performed Cangaroo::ProductPollJob from Inline(cangaroo) in 3745.46ms
Cangaroo::Webhook::Error: The property '#/' contains additional properties ["request_id"] outside of the schema when none are allowed in schema Cangaroo Object#
	from /Users/bci20/Sites/acc-001-api/.gems/bundler/gems/cangaroo-9b097351a784/app/jobs/cangaroo/poll_job.rb:38:in `perform'
```

**Comment by bricesanchez** [comment]
I'm trying to know how could i post my data to another API after my polling.

For example : 
I've Chaindrive as an ERP with a SOAP API and Shopify as my ecommerce platform with a REST API.
I poll products data from Chaindrive and want to post them to Shopify.
This is my current ProductPollJob : 
```ruby
module Cangaroo
  class ProductPollJob < Cangaroo::PollJob
    connection 'chaindrive'
    frequency 1.hour
    path '/get_products'
    parameters(
      { 
        product_filters: {
          item: 'PRODUCTID=270112'
        }
      }
    )
  end
end
```

Is there something like a destination method ?

**Comment by iloveitaly** [comment]
@bricesanchez you'd create another job to indicate how cangaroo should process the ingested payloads from the polling URL. Checkout this line: https://github.com/nebulab/cangaroo/pull/11/files#diff-f894b9c7f5c707cf739ec28b47435851R19

---

@AlessioRocco Could you review these PRs first, before review this one:

* https://github.com/nebulab/cangaroo/pull/21
* https://github.com/nebulab/cangaroo/pull/20
* https://github.com/nebulab/cangaroo/pull/18

This will make it easier for me to update this PR with the latest changes I've made to this PR. Thanks!

**Comment by bricesanchez** [comment]
Thanks @iloveitaly, i will test your answer.

Do you have any idea for my previous comment ? : https://github.com/nebulab/cangaroo/pull/11#issuecomment-214505913

**Comment by iloveitaly** [comment]
@bricesanchez It's a JSON schema validation issue. I've run into this with other integrations as well. A quick hack is to [remove the schema validator from the interactor chain](https://github.com/nebulab/cangaroo/blob/20693a79b6620a1eb7a14815e22999dc44c8d433/app/interactors/cangaroo/perform_flow.rb#L5) or [edit the schema to match your incoming payload](https://github.com/nebulab/cangaroo/blob/20693a79b6620a1eb7a14815e22999dc44c8d433/app/interactors/cangaroo/validate_json_schema.rb#L5).

**Comment by bricesanchez** [comment]
@iloveitaly Nice thank you i test now :)

**Comment by rohitnick** [comment]
@bricesanchez @iloveitaly https://github.com/nebulab/cangaroo/pull/11#issuecomment-214505913 can be solved with pr #23 , please have a look. I have raised an issue #22 for the same error.

If you find the fix PR ok, then I can go ahead and make the test cases pass.

**Comment by bricesanchez** [comment]
> @bricesanchez you'd create another job to indicate how cangaroo should process the ingested payloads from the polling URL. Checkout this line: https://github.com/nebulab/cangaroo/pull/11/files#diff-f894b9c7f5c707cf739ec28b47435851R19

I'm confused on how to make it work, could you help me ?

**Comment by iloveitaly** [comment]
@bricesanchez Here's an example:

```ruby
require 'rollbar/active_job'

module Spree
  class AddCustomerJob < Cangaroo::Job
    connection :spree
    path '/api/pipeline/users'

    def perform?
      type == 'customers' &&
      self.destination_connection != self.source_connection
    end
  end
end
```

Sorry I can't give more specifics, short on time.

**Comment by bricesanchez** [comment]
@iloveitaly Thanks it's what i'm doing on the Job but it return an error like `perform? is NotYetImplemented`even if i've this method in my Job :(

**Comment by iloveitaly** [comment]
@bricesanchez `perform?` has to be implemented on the job subclass. Can you post your job subclass here?

**Comment by bricesanchez** [comment]
```ruby
module Cangaroo
  class ProductJob < Cangaroo::Job
    connection 'shopify'
    path '/add_product'

    def transform
       payload
    end

    def perform?
      type = 'products'
    end
  end
end
```

This should works?

**Comment by AlessioRocco** [comment]
@bricesanchez `perform?` should return `true` or `false`, I'm not sure it's the only problem but try to change it for now.

**Comment by bricesanchez** [comment]
One PR remaining ( #20 ) until you can merge this PR 😀

**Comment by iloveitaly** [comment]
@AlessioRocco Rebased and updated with latest tweaks. We are good to go here!

**Comment by AlessioRocco** [comment]
@iloveitaly finally merged, thanks!
